src/modules/bookings/bookings.service.ts

```diff
@@ -0,0 +1,190 @@
+  ) {
+    const queryBuilder = this.bookingRepository.createQueryBuilder('booking');
+
+    queryBuilder.where('booking.userId = :userId', { userId });
+
+    if (queryParamsDto.status) {
+      queryBuilder.andWhere('booking.status = :status', {
+        status: queryParamsDto.status,
+      });
+    }
+
+    queryBuilder
+      .leftJoinAndSelect('booking.accommodation', 'accommodation')
+      .select(['booking', 'accommodation']);
+
+    return await queryBuilder.getMany();
+  }
+
+  async getBooking(userId: string, bookingId: string): Promise<Booking> {
+    const booking = await this.bookingRepository.findOne({
+      where: { id: bookingId },
+      relations: ['accommodation', 'user', 'accommodation.owner'],
+    });
+
+    if (
+      booking.accommodation.owner.id !== userId &&
+      booking.user.id !== userId
+    ) {
+      throw new UnauthorizedException(
+        'You are not allowed to view this booking',
+      );
+    }
+
+    if (!booking) {
+      throw new NotFoundException(`Booking with id ${bookingId} not found`);
+    }
+
+    return booking;
+  }
+
+  async createBooking(
+    dto: CreateBookingReqDto,
+    userId: string,
+  ): Promise<Booking> {
+    const accommodation = await this.accommodationsService.getAccommodationById(
+      dto.accommodationId,
+    );
+
+    if (!accommodation) {
+      throw new NotFoundException(
+        `Accommodation with id ${dto.accommodationId} not found`,
+      );
+    }
+
+    if (accommodation.owner.id === userId) {
+      throw new BadRequestException("You can't book your own accommodation");
+    }
+
+    if (dto.checkOutDate <= dto.checkInDate) {
+      throw new BadRequestException('Check out date should be after check in');
+    }
+
+    if (dto.checkInDate <= new Date()) {
+      throw new BadRequestException("Check in date can't be in the past");
+    }
+
+    const existingBookings = await this.bookingRepository
+      .createQueryBuilder('booking')
+      .where('booking.accommodationId = :accommodationId', {
+        accommodationId: dto.accommodationId,
+      })
+      .andWhere(
+        `booking.checkInDate <= :checkInDate AND booking.checkOutDate >= :checkInDate OR 
+         booking.checkInDate <= :checkOutDate AND booking.checkOutDate >= :checkOutDate OR
+         booking.checkInDate >= :checkInDate AND booking.checkOutDate <= :checkOutDate`,
+        { checkInDate: dto.checkInDate, checkOutDate: dto.checkOutDate },
+      )
+      .andWhere('booking.status IN (:...activeStatuses)', {
+        activeStatuses: [BookingStatus.CONFIRMED, BookingStatus.CHECKED_IN],
+      })
+      .getCount();
+
+    if (existingBookings > 0) {
+      throw new BadRequestException(
+        'Accommodation is already booked for the selected dates',
+      );
+    }
+
+    const nights = this.calculateNights(dto.checkInDate, dto.checkOutDate);
+
+    const newBooking = this.bookingRepository.create({
+      checkInDate: dto.checkInDate,
+      checkOutDate: dto.checkOutDate,
```

Comment:
Format dates. Just in case.

Reply:
I moved date formatting logic to the dto validation layer